design/animation.md

```diff
@@ -0,0 +1,317 @@
+argument. Over the wire these are represented as strings, but are easily generated in the API with helper functions.
+
+```ts
+class TargetPath<T> {
+	private parent: TargetPath<any>;
+	constructor(name: string, children: Map<string, TargetPath<any>>) {
+		for (const [k,v] of children) {
+			this[k] = v;
+			v.parent = this;
+		}
+	}
+	public toJSON() {
+		// convert tree into string
+	}
+}
+function AnimateActor(index: number) {
+	return new TargetPath<never>(`actor.${index}`, {
+		transform: new TargetPath<never>("transform", {
+			local: new TargetPath<never>("local", {
+				position: new TargetPath<Vector3>("position", {
+					x: new TargetPath<number>("x"),
+					y: new TargetPath<number>("y"),
+					z: new TargetPath<number>("z")
+				})
+			})
+		})
+	});
+}
+
+const track = {
+	target: AnimateActor(0).transform.local.position
+	// ...
+} as Track<Vector3>;
+```
+
+### Instances
+
+Animation instances are animations that have been bound to a particular set of actors and materials. These instances
+can be created by explicitly binding an animation (`Animation.bind(...objects)`), or may be returned during actor
+creation from `Actor.CreateFromPrefab`/`Gltf`/`Library`. Unlike the asset, instances can be played, paused, set speed,
+etc. Animation instances can either be stored directly, or obtained from any of the actors/materials that participate
+in the animation. Instances participate in the patching system, so changing properties on the animation instance
+are pushed down to clients automatically.
+
+An animation instance can be set to start another animation on the instant that it finishes, by setting an instance's
+`continueWith` property. Because it's set in advance, there is zero downtime between one ending and another starting,
+and does not require a round trip.
+
+```ts
+class Animation {
+	bind(...args: Animatible[]): AnimationInstance;
+}
+
+class AnimationInstance {
+	/** Generated unique ID */
+	id: Readonly<string>;
+
+	/** Playback speed multiplier */
+	speed: number;
+	/** What happens when the animation hits the last frame */
+	wrapMode: WrapMode;
+	/** When multiple animations play together, this is the relative strength of this instance */
+	weight: number;
+	/** The current playback time, based on start time and speed */
+	time: number;
+	/** The list of actors/materials that participate in this instance */
+	targets: Readonly<Animatible[]>;
+	/** The playing animation */
+	animation: Readonly<Animation>;
+	/** Start these animation instances after this one completes */
+	continueWith: AnimationInstance[];
+
+	/** Current play state */
+	isPlaying: Readonly<boolean>;
+
+	play(): void;
+	pause(): void;
+
+	/** Only fired if wrapMode is Once */
+	finished(): Promise<void>;
+	/** Sets the `continueWith` property, and returns the next anim instance */
+	continueWith(otherAnim: AnimationInstace): AnimationInstance;
```

Comment:
`continueWith` is defined twice in this interface.

Reply:
`continueWith` is specialized functionality that should live outside this class. I'd put this functionality into a new class called `AnimationGroup`, or `AnimationChain`. It would consist of a set of `AnimationInstance`s and rules for how to play them together. Like a command buffer for animations.

Reply:
I thought a linked-list type chain like this would be simpler and easier to understand. I'm curious about what you have in mind though, any more details?

---
In reply to: [341702219](https://github.com/microsoft/mixed-reality-extension-sdk/pull/464#discussion_r341702219) [](ancestors = 341702219)